src/main/java/co/cask/common/AvroToStructuredTransformer.java

```diff
@@ -0,0 +1,148 @@
+ * the License.
+ */
+
+package co.cask.common;
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.format.UnexpectedFormatException;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.hydrator.common.RecordConverter;
+import com.google.common.collect.Maps;
+import org.apache.avro.generic.GenericRecord;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Create StructuredRecords from GenericRecords
+ */
+public class AvroToStructuredTransformer extends RecordConverter<GenericRecord, StructuredRecord> {
+  private final Map<Integer, Schema> schemaCache = Maps.newHashMap();
+
+  public StructuredRecord transform(GenericRecord genericRecord) throws IOException {
+    org.apache.avro.Schema genericRecordSchema = genericRecord.getSchema();
+    return transform(genericRecord, convertSchema(genericRecordSchema));
+  }
+
+  @Override
+  public StructuredRecord transform(GenericRecord genericRecord, Schema structuredSchema) throws IOException {
+    StructuredRecord.Builder builder = StructuredRecord.builder(structuredSchema);
+    for (Schema.Field field : structuredSchema.getFields()) {
+      String fieldName = field.getName();
+      Object value = convertField(genericRecord.get(fieldName), field.getSchema());
+      if (value instanceof LocalDate) {
+        builder.setDate(fieldName, (LocalDate) value);
+      } else if (value instanceof LocalTime) {
+        builder.setTime(fieldName, (LocalTime) value);
+      } else if (value instanceof ZonedDateTime) {
+        builder.setTimestamp(fieldName, (ZonedDateTime) value);
+      } else {
+        builder.set(fieldName, value);
+      }
+    }
+    return builder.build();
+  }
+
+  @Override
+  @Nullable
+  protected Object convertField(Object field, Schema fieldSchema) throws IOException {
+    if (field == null) {
+      return null;
+    }
+
+    Schema.Type fieldType = fieldSchema.getType();
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+
+    if (logicalType != null) {
+      switch (logicalType) {
+        case DATE:
+          // date will be in yyyy-mm-dd format
+          return LocalDate.parse(field.toString());
+        case TIME_MILLIS:
+        case TIME_MICROS:
+          // time will be in hh:mm:ss format
+          return LocalTime.parse(field.toString());
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          return getZonedDateTime((long) field);
+        default:
+          throw new UnexpectedFormatException("field type " + fieldType + " is not supported.");
+      }
+    }
+
+    // This casting is added because bigquery up casts number fields. For example:
+    // int becomes long, float becomes double
```

Comment:
this seems really weird... are we sure it always happens?

Reply:
Its documented here: https://cloud.google.com/bigquery/docs/loading-data-cloud-storage-avro. 
Also I verified it by having a bi query table with INTEGER and FLOAT types in schema.

Reply:
That doc is for loading data into big query, not reading it.

Reply:
Its both ways, for export and import: https://cloud.google.com/bigquery/docs/exporting-data#avro_export_details 

![image](https://user-images.githubusercontent.com/14131070/44611130-61b4f080-a7b4-11e8-8b7f-6301f769ea3c.png)

For example for a table with below schema, avro schema looks as below. Please note that `unique_key` is interpreted as `long` in avro schema and `latitude` is interpreted as `double`

![image](https://user-images.githubusercontent.com/14131070/44611211-c96b3b80-a7b4-11e8-89f9-fbef106a2eb9.png)



```json
{
  "type": "record",
  "name": "Root",
  "fields": [
    {
      "name": "unique_key",
      "type": [
        "null",
        "long"
      ]
    },
    {
      "name": "descript",
      "type": [
        "null",
        "string"
      ]
    },
    {
      "name": "date",
      "type": [
        "null",
        "string"
      ]
    },
    {
      "name": "time",
      "type": [
        "null",
        "string"
      ]
    },
    {
      "name": "address",
      "type": [
        "null",
        "string"
      ]
    },
    {
      "name": "longitude",
      "type": [
        "null",
        "double"
      ]
    },
    {
      "name": "latitude",
      "type": [
        "null",
        "double"
      ]
    },
    {
      "name": "location",
      "type": [
        "null",
        "string"
      ]
    },
    {
      "name": "timestamp",
      "type": [
        "null",
        {
          "type": "long",
          "logicalType": "timestamp-micros"
        }
      ]
    }
  ]
}
```

Reply:
Looked at the docs some more and this is because there is no 'int' and 'long' in bigquery, and no 'float' and 'double'. A big query 'integer' is 8 bytes, which is a java 'long', and a big query 'float' is 8 bytes, which is a java double. So this actually means the avro int and float types should never be returned by the getSchema call